zdm/grid.py

```diff
@@ -628,26 +656,41 @@ def update(self, vparams:dict, ALL=False):
         # TODO -- For cubes with multiple surveys can we do these
         #   first two steps (even the first 5!) only once??
         # Update cosmology?
-        if reset_cos:
+        if reset_cos and prev_grid is None:
             cos.set_cosmology(self.state)
             cos.init_dist_measures()
 
         if get_zdm or ALL:
-            zDMgrid, zvals,dmvals=misc_functions.get_zdm_grid(
-                self.state, new=True,plot=False,method='analytic')
-            # TODO -- Check zvals and dmvals haven't changed!
-            self.pass_grid(zDMgrid,zvals,dmvals)
+            if prev_grid is None:
+                zDMgrid, zvals,dmvals=misc_functions.get_zdm_grid(
+                    self.state, new=True,plot=False,method='analytic')
+                self.pass_grid(zDMgrid,zvals,dmvals)
+            else:
+                # Pass a copy (just to be safe)
+                self.pass_grid(prev_grid.grid.copy(),
+                               prev_grid.zvals.copy(),
+                               prev_grid.dmvals.copy())
 
         if calc_dV or ALL:
-            self.calc_dV()
+            if prev_grid is None:
+                self.calc_dV()
+            else:
+                self.dV = prev_grid.dV.copy()
 
         # Smear?
         if smear_mask or ALL:
-            self.smear=pcosmic.get_dm_mask(
-                self.dmvals,(self.state.host.lmean,
+            if prev_grid is None:
+                self.smear=pcosmic.get_dm_mask(
+                    self.dmvals,(self.state.host.lmean,
                              self.state.host.lsigma), self.zvals)
+            else:
+                self.smear = prev_grid.smear.copy()
         if smear_dm or ALL:
-            self.smear_dm(self.smear)
+            if prev_grid is None:
+                self.smear_dm(self.smear)
+            else:
+                self.smear = prev_grid.smear.copy()
+                self.smear_grid = prev_grid.smear_grid.copy()
```

Comment:
OK, I really like this ability to just copy the necessary info, it should make it much easier to use multiple grids. We might want to chase up in future whether or not to use "copy" here, since a bunch of these should remain untouched once created, and the difference could be significant for very many FRBs - but we're not at that point yet.